chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/ics/reviewer/MissingIcsSurveillanceReviewer.java

```diff
@@ -0,0 +1,101 @@
+import gov.healthit.chpl.domain.surveillance.Surveillance;
+import gov.healthit.chpl.dto.CertifiedProductDTO;
+import gov.healthit.chpl.manager.SurveillanceManager;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2(topic = "icsErrorsReportCreatorJobLogger")
+@Component
+public class MissingIcsSurveillanceReviewer extends IcsErrorsReviewer {
+    private static final int MAX_INHERITED_GENERATIONS_WITHOUT_SURVEILLANCE = 3;
+    private static final String ICS_REQUIREMENT_TYPE = "Inherited Certified Status";
+    private SurveillanceManager survManager;
+    private ListingGraphDAO listingGraphDao;
+    private String errorMessage;
+
+    @Autowired
+    public MissingIcsSurveillanceReviewer(SurveillanceManager survManager,
+            ListingGraphDAO listingGraphDao,
+            @Value("${ics.missingSurveillanceError}") String errorMessage) {
+        this.survManager = survManager;
+        this.listingGraphDao = listingGraphDao;
+        this.errorMessage = errorMessage;
+    }
+
+    @Override
+    public String getIcsError(CertifiedProductSearchDetails listing) {
+        //it's an error if none of the Listing's "self", "parent",
+        //"grandparent", "great-grandparent" has surveillance of requirement type "Inherited Certified Status"
+        if (hasIcs(listing)) {
+            int generationsWithoutIcsSurveillanceCount =
+                    getGenerationsWithoutIcsSurveillanceCount(Stream.of(listing.getId()).collect(Collectors.toList()),
+                            new ArrayList<Long>(),
+                            Boolean.FALSE, 0);
+            if (generationsWithoutIcsSurveillanceCount > MAX_INHERITED_GENERATIONS_WITHOUT_SURVEILLANCE) {
+                return errorMessage;
+            }
+        }
+        return null;
+    }
+
+    private Integer getGenerationsWithoutIcsSurveillanceCount(List<Long> listingIds, List<Long> checkedListingIds, Boolean foundIcsSurveillance, Integer generationsWithoutIcsSurveillanceCount) {
+        Iterator<Long> listingIdIter = listingIds.iterator();
+        while (!foundIcsSurveillance && listingIdIter.hasNext()) {
+            Long listingId = listingIdIter.next();
+            if (checkedListingIds.contains(listingId)) {
+                continue;
+            }
+
+            checkedListingIds.add(listingId);
+            List<Surveillance> surveillances = survManager.getByCertifiedProduct(listingId);
+            boolean hasIcsSurveillance = isAnySurveillanceForIcs(surveillances);
+            if (!hasIcsSurveillance) {
+                LOGGER.debug("\tNo ICS Surveillance found for listing: " + listingId);
+                List<CertifiedProductDTO> parents = getParents(listingId);
+                if (CollectionUtils.isEmpty(parents)) {
+                    return generationsWithoutIcsSurveillanceCount;
+                } else {
+                    return getGenerationsWithoutIcsSurveillanceCount(parents.stream().map(parent -> parent.getId()).toList(),
```

Comment:
I was wondering if this could be recursive. I'm intrigued to see it

Reply:
It wasn't the first time I wrote it, but it really should be so I re-wrote and I added enough unit tests that I'm pretty sure it's working.